rasa/nlu/extractors/crf_role_extractor.py

```diff
@@ -0,0 +1,575 @@
+                "instructions."
+            )
+
+    def _validate_configuration(self):
+        if len(self.component_config.get("features", [])) % 2 != 1:
+            raise ValueError(
+                "Need an odd number of crf feature lists to have a center word."
+            )
+
+    @classmethod
+    def required_packages(cls):
+        return ["sklearn_crfsuite", "sklearn"]
+
+    def train(
+        self, training_data: TrainingData, config: RasaNLUModelConfig, **kwargs: Any
+    ) -> None:
+
+        # checks whether there is at least one
+        # example with an entity annotation
+        if training_data.entity_examples:
+            self._check_spacy_doc(training_data.training_examples[0])
+
+            # do not filter out pre-trained entities since we can still train a role for them
+            # convert entity examples to role examples
+            role_examples = self.create_role_examples(training_data.entity_examples)
+
+            # convert the dataset into features
+            # this will train on ALL examples, even the ones
+            # without annotations
+            dataset = self._create_dataset(role_examples)
+
+            self._train_model(dataset)
+
+
+    def _create_dataset(
+        self, examples: List[Message]
+    ) -> List[List[Tuple[Optional[Text], Optional[Text], Text, Dict[Text, Any]]]]:
+        dataset = []
+        for example in examples:
+            entity_offsets = self._convert_example(example)
+            dataset.append(self._from_json_to_crf(example, entity_offsets))
+        return dataset
+
+    def _check_spacy_doc(self, message):
+        if self.pos_features and message.get("spacy_doc") is None:
+            raise InvalidConfigError(
+                "Could not find `spacy_doc` attribute for "
+                "message {}\n"
+                "POS features require a pipeline component "
+                "that provides `spacy_doc` attributes, i.e. `SpacyNLP`. "
+                "See https://nlu.rasa.com/pipeline.html#nlp-spacy "
+                "for details".format(message.text)
+            )
+
+    def process(self, message: Message, **kwargs: Any) -> None:
+
+        self._check_spacy_doc(message)
+
+        extracted = self.extract_roles(message)
+        #extracted = self.add_extractor_name(self.extract_roles(message))
+        message.set(
+            "roles", message.get("roles", []) + extracted, add_to_output=True
+        )
+
+    @staticmethod
+    def _convert_example(example: Message) -> List[Tuple[int, int, Text]]:
+        def convert_entity(entity):
+            return entity["start"], entity["end"], entity["entity"]
+
+        return [convert_entity(ent) for ent in example.get("entities", [])]
+
+    def extract_roles(self, message: Message) -> List[Dict[Text, Any]]:
+        """Take a sentence and return roles in json format"""
+
+        if self.ent_tagger is not None:            
+            role_message = self.replace_entities_with_roles(message)
```

Comment:
I guess at prediction time this will not work. We want to have a sentence like `heading for location and flying out of location` in the end. However, the function `replace_entites_with_roles` would do `heading for [location](destination) and flying out of [location](origin)`. It is not replacing the actual text, is it? Or do I miss something?